Support Elasticsearch 7

Depends on https://github.com/jaegertracing/jaeger/pull/1690

With ES 7 it fails on 
```
docker run --env STORAGE=elasticsearch --env ES_NODES=http://localhost:9200  --net=host jaegertracing/spark-dependencies              3:23 

19/08/13 13:23:04 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
19/08/13 13:23:05 INFO ElasticsearchDependenciesJob: Running Dependencies job for 2019-08-13T00:00Z, reading from jaeger-span-2019-08-13 index, result storing to jaeger-dependencies-2019-08-13
Exception in thread "main" org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: Cannot detect ES version - typically this happens if the network/Elasticsearch cluster is not accessible or when targeting a WAN/Cloud instance without the proper setting 'es.nodes.wan.only'
	at org.elasticsearch.hadoop.rest.InitializationUtils.discoverEsVersion(InitializationUtils.java:327)
	at org.elasticsearch.hadoop.rest.RestService.findPartitions(RestService.java:220)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.esPartitions$lzycompute(AbstractEsRDD.scala:73)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.esPartitions(AbstractEsRDD.scala:72)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.getPartitions(AbstractEsRDD.scala:44)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.Partitioner$$anonfun$defaultPartitioner$2.apply(Partitioner.scala:66)
	at org.apache.spark.Partitioner$$anonfun$defaultPartitioner$2.apply(Partitioner.scala:66)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at org.apache.spark.Partitioner$.defaultPartitioner(Partitioner.scala:66)
	at org.apache.spark.rdd.RDD$$anonfun$groupBy$1.apply(RDD.scala:688)
	at org.apache.spark.rdd.RDD$$anonfun$groupBy$1.apply(RDD.scala:688)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
	at org.apache.spark.rdd.RDD.groupBy(RDD.scala:687)
	at org.apache.spark.api.java.JavaRDDLike$class.groupBy(JavaRDDLike.scala:243)
	at org.apache.spark.api.java.AbstractJavaRDDLike.groupBy(JavaRDDLike.scala:45)
	at io.jaegertracing.spark.dependencies.elastic.ElasticsearchDependenciesJob.run(ElasticsearchDependenciesJob.java:193)
	at io.jaegertracing.spark.dependencies.elastic.ElasticsearchDependenciesJob.run(ElasticsearchDependenciesJob.java:172)
	at io.jaegertracing.spark.dependencies.DependenciesSparkJob.run(DependenciesSparkJob.java:54)
	at io.jaegertracing.spark.dependencies.DependenciesSparkJob.main(DependenciesSparkJob.java:40)
Caused by: org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: Unsupported/Unknown Elasticsearch version 7.3.0
	at org.elasticsearch.hadoop.util.EsMajorVersion.parse(EsMajorVersion.java:82)
	at org.elasticsearch.hadoop.rest.RestClient.remoteEsVersion(RestClient.java:664)
	at org.elasticsearch.hadoop.rest.InitializationUtils.discoverEsVersion(InitializationUtils.java:320)
	... 33 more

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -29,7 +29,7 @@ jobs:
 
     - stage: test
       env: [ NAME=Jaeger latest, ES7 ]
-      script: ./mvnw clean test
+      script: ELASTICSEARCH_VERSION=7.3.0 ./mvnw clean test -pl jaeger-spark-dependencies-elasticsearch
 
 # Don't build release tags. This avoids publish conflicts because the version commit exists both on master and the release tag.
 # See https://github.com/travis-ci/travis-ci/issues/1532

--- a/.travis.yml
+++ b/.travis.yml
@@ -31,10 +31,6 @@ jobs:
       env: [ NAME=Jaeger latest, ES7 ]
       script: ./mvnw clean test
 
-    - stage: test
-      env: [ NAME=Jaeger latest, ES5 ]
-      script: ELASTICSEARCH_VERSION=7.3.0 ./mvnw clean test -pl jaeger-spark-dependencies-elasticsearch
-
 # Don't build release tags. This avoids publish conflicts because the version commit exists both on master and the release tag.
 # See https://github.com/travis-ci/travis-ci/issues/1532
 branches:

--- a/.travis.yml
+++ b/.travis.yml
@@ -27,6 +27,13 @@ jobs:
       env: [ NAME=Jaeger latest ]
       script: ./mvnw clean test
 
+    - stage: test
+      env: [ NAME=Jaeger latest, ES7 ]
+      script: ./mvnw clean test
+
+    - stage: test
+      env: [ NAME=Jaeger latest, ES5 ]
+      script: ELASTICSEARCH_VERSION=7.3.0 ./mvnw clean test -pl jaeger-spark-dependencies-elasticsearch
 
 # Don't build release tags. This avoids publish conflicts because the version commit exists both on master and the release tag.
 # See https://github.com/travis-ci/travis-ci/issues/1532

--- a/README.md
+++ b/README.md
@@ -89,6 +89,7 @@ $ STORAGE=elasticsearch ES_NODES=http://localhost:9200 java -jar jaeger-spark-de
 To build the job locally and run tests:
 ```bash
 ./mvnw clean install # if failed add SPARK_LOCAL_IP=127.0.0.1
+STORAGE=elasticsearch ES_NODES=http://localhost:9200 java -jar jaeger-spark-dependencies/target/jaeger-spark-dependencies-0.0.1-SNAPSHOT.jar
 docker build -t jaegertracing/spark-dependencies:latest .
 ```
 

--- a/jaeger-spark-dependencies-cassandra/pom.xml
+++ b/jaeger-spark-dependencies-cassandra/pom.xml
@@ -37,13 +37,6 @@
     <dependency>
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-core_${version.scala.binary}</artifactId>
-      <exclusions>
-        <!-- exclude netty-all as it interferes with cassandra's netty -->
-        <exclusion>
-          <groupId>io.netty</groupId>
-          <artifactId>netty-all</artifactId>
-        </exclusion>
-      </exclusions>
     </dependency>
 
     <dependency>

--- a/jaeger-spark-dependencies-cassandra/pom.xml
+++ b/jaeger-spark-dependencies-cassandra/pom.xml
@@ -25,7 +25,7 @@
   <artifactId>jaeger-spark-dependencies-cassandra</artifactId>
 
   <properties>
-    <spark-cassandra-connector.version>2.0.3</spark-cassandra-connector.version>
+    <spark-cassandra-connector.version>2.3.3</spark-cassandra-connector.version>
   </properties>
 
   <dependencies>

--- a/jaeger-spark-dependencies-cassandra/src/test/java/io/jaegertracing/spark/dependencies/cassandra/CassandraDependenciesJobTest.java
+++ b/jaeger-spark-dependencies-cassandra/src/test/java/io/jaegertracing/spark/dependencies/cassandra/CassandraDependenciesJobTest.java
@@ -47,7 +47,7 @@ public class CassandraDependenciesJobTest extends DependenciesTest {
   @BeforeClass
   public static void beforeClass() throws TimeoutException {
     network = Network.newNetwork();
-    cassandra = new CassandraContainer("cassandra:3.9")
+    cassandra = new CassandraContainer("cassandra:3.11")
         .withNetwork(network)
         .withNetworkAliases("cassandra")
         .withExposedPorts(9042);
@@ -68,7 +68,8 @@ public class CassandraDependenciesJobTest extends DependenciesTest {
         .withEnv("CASSANDRA_KEYSPACE", "jaeger_v1_dc1")
         .withEnv("COLLECTOR_ZIPKIN_HTTP_PORT", "9411")
         .withEnv("COLLECTOR_QUEUE_SIZE", "100000")
-        .waitingFor(new BoundPortHttpWaitStrategy(14269).forStatusCode(204))
+        // older versions of jaeger were using 204 status code, now changed to 200
+        .waitingFor(new BoundPortHttpWaitStrategy(14269).forStatusCodeMatching(statusCode -> statusCode >= 200 && statusCode < 300))
         // the first one is health check
         .withExposedPorts(14269, 14268, 9411);
     jaegerCollector.start();
@@ -77,9 +78,7 @@ public class CassandraDependenciesJobTest extends DependenciesTest {
         .withNetwork(network)
         .withEnv("CASSANDRA_SERVERS", "cassandra")
         .withEnv("CASSANDRA_KEYSPACE", "jaeger_v1_dc1")
-        // TODO remove after https://github.com/jaegertracing/jaeger/pull/1364 is merged
-        .withEnv("CASSANDRA_ENABLE_DEPENDENCIES_V2", "true")
-        .waitingFor(new BoundPortHttpWaitStrategy(16687).forStatusCode(204))
+        .waitingFor(new BoundPortHttpWaitStrategy(16687).forStatusCodeMatching(statusCode -> statusCode >= 200 && statusCode < 300))
         .withExposedPorts(16687, 16686);
     jaegerQuery.start();
 

--- a/jaeger-spark-dependencies-cassandra/src/test/java/io/jaegertracing/spark/dependencies/cassandra/CassandraDependenciesJobTest.java
+++ b/jaeger-spark-dependencies-cassandra/src/test/java/io/jaegertracing/spark/dependencies/cassandra/CassandraDependenciesJobTest.java
@@ -16,7 +16,6 @@ package io.jaegertracing.spark.dependencies.cassandra;
 import static org.awaitility.Awaitility.await;
 
 import com.datastax.driver.core.Cluster;
-import com.datastax.driver.core.Row;
 import com.datastax.driver.core.Session;
 import io.jaegertracing.spark.dependencies.test.DependenciesTest;
 import java.time.LocalDate;
@@ -24,7 +23,6 @@ import java.util.Collections;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
 import org.junit.After;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
@@ -45,7 +43,7 @@ public class CassandraDependenciesJobTest extends DependenciesTest {
   private static int cassandraPort;
 
   @BeforeClass
-  public static void beforeClass() throws TimeoutException {
+  public static void beforeClass() {
     network = Network.newNetwork();
     cassandra = new CassandraContainer("cassandra:3.11")
         .withNetwork(network)
@@ -113,7 +111,7 @@ public class CassandraDependenciesJobTest extends DependenciesTest {
   }
 
   @Override
-  protected void deriveDependencies() throws Exception {
+  protected void deriveDependencies() {
     CassandraDependenciesJob.builder()
         .contactPoints("localhost:" + cassandraPort)
         .day(LocalDate.now())

--- a/jaeger-spark-dependencies-elasticsearch/pom.xml
+++ b/jaeger-spark-dependencies-elasticsearch/pom.xml
@@ -35,7 +35,7 @@
     <dependency>
       <groupId>org.elasticsearch</groupId>
       <artifactId>elasticsearch-spark-20_${version.scala.binary}</artifactId>
-      <version>6.0.0-beta2</version>
+      <version>7.3.0</version>
     </dependency>
 
     <dependency>

--- a/jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java
@@ -192,6 +192,7 @@ public class ElasticsearchDependenciesJob {
             .map(new ElasticTupleToSpan())
             .groupBy(Span::getTraceId);
         List<Dependency> dependencyLinks = DependenciesSparkHelper.derive(traces,peerServiceTag);
+        // TODO the type names are deprecated (in ES7) and will be removed in a later release.
         store(sc, dependencyLinks, depIndex + "/dependencies");
         log.info("Done, {} dependency objects created", dependencyLinks.size());
         if (dependencyLinks.size() > 0) {

--- a/jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java
@@ -31,6 +31,9 @@ import java.util.Map;
 import org.apache.spark.SparkConf;
 import org.apache.spark.api.java.JavaPairRDD;
 import org.apache.spark.api.java.JavaSparkContext;
+import org.elasticsearch.hadoop.rest.RestClient;
+import org.elasticsearch.hadoop.util.EsMajorVersion;
+import org.elasticsearch.spark.cfg.SparkSettings;
 import org.elasticsearch.spark.rdd.api.java.JavaEsSpark;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -192,8 +195,14 @@ public class ElasticsearchDependenciesJob {
             .map(new ElasticTupleToSpan())
             .groupBy(Span::getTraceId);
         List<Dependency> dependencyLinks = DependenciesSparkHelper.derive(traces,peerServiceTag);
-        // TODO the type names are deprecated (in ES7) and will be removed in a later release.
-        store(sc, dependencyLinks, depIndex + "/dependencies");
+        EsMajorVersion esMajorVersion = getEsVersion();
+        // Add type for ES < 7
+        // WARN log is produced for older ES versions, however it's produced by spark-es library and not ES itself, it cannot be disabled
+        //  WARN Resource: Detected type name in resource [jaeger-dependencies-2019-08-14/dependencies]. Type names are deprecated and will be removed in a later release.
+        if (esMajorVersion.before(EsMajorVersion.V_7_X)) {
+          depIndex = depIndex + "/dependencies";
+        }
+        store(sc, dependencyLinks, depIndex);
         log.info("Done, {} dependency objects created", dependencyLinks.size());
         if (dependencyLinks.size() > 0) {
           // we do not derive dependencies for old prefix "prefix:" if new prefix "prefix-" contains data
@@ -202,6 +211,15 @@ public class ElasticsearchDependenciesJob {
       }
     } finally {
       sc.stop();
+    }
+  }
+
+  private EsMajorVersion getEsVersion() {
+    RestClient client = new RestClient(new SparkSettings(conf));
+    try {
+      return client.mainInfo().getMajorVersion();
+    } finally {
+      client.close();
     }
   }
 

--- a/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJobTest.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJobTest.java
@@ -41,7 +41,7 @@ public class ElasticsearchDependenciesJobTest extends DependenciesTest {
   @BeforeClass
   public static void beforeClass() {
     jaegerElasticsearchEnvironment = new JaegerElasticsearchEnvironment();
-    jaegerElasticsearchEnvironment.start(new HashMap<>(), jaegerVersion());
+    jaegerElasticsearchEnvironment.start(new HashMap<>(), jaegerVersion(), JaegerElasticsearchEnvironment.elasticsearchVersion());
     collectorUrl = jaegerElasticsearchEnvironment.getCollectorUrl();
     zipkinCollectorUrl = jaegerElasticsearchEnvironment.getZipkinCollectorUrl();
     queryUrl = jaegerElasticsearchEnvironment.getQueryUrl();

--- a/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesTagFieldsJobTest.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesTagFieldsJobTest.java
@@ -27,7 +27,7 @@ public class ElasticsearchDependenciesTagFieldsJobTest extends ElasticsearchDepe
     jaegerElasticsearchEnvironment = new JaegerElasticsearchEnvironment();
     HashMap<String, String> jaegerEnvSetting = new HashMap<>();
     jaegerEnvSetting.put("ES_TAGS__AS_FIELDS_ALL", "true");
-    jaegerElasticsearchEnvironment.start(jaegerEnvSetting, jaegerVersion());
+    jaegerElasticsearchEnvironment.start(jaegerEnvSetting, jaegerVersion(), JaegerElasticsearchEnvironment.elasticsearchVersion());
     collectorUrl = jaegerElasticsearchEnvironment.getCollectorUrl();
     zipkinCollectorUrl = jaegerElasticsearchEnvironment.getZipkinCollectorUrl();
     queryUrl = jaegerElasticsearchEnvironment.getQueryUrl();

--- a/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/JaegerElasticsearchEnvironment.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/JaegerElasticsearchEnvironment.java
@@ -72,7 +72,7 @@ public class JaegerElasticsearchEnvironment {
         .withEnv("COLLECTOR_ZIPKIN_HTTP_PORT", "9411")
         .withEnv("COLLECTOR_QUEUE_SIZE", "100000")
         .withEnv(jaegerEnvs)
-        .waitingFor(new BoundPortHttpWaitStrategy(14269).forStatusCode(204))
+        .waitingFor(new BoundPortHttpWaitStrategy(14269).forStatusCodeMatching(statusCode -> statusCode >= 200 && statusCode < 300))
         // the first one is health check
         .withExposedPorts(14269, 14268, 9411);
     jaegerCollector.start();
@@ -83,7 +83,7 @@ public class JaegerElasticsearchEnvironment {
         .withEnv("ES_TAGS_AS_FIELDS_ALL", "true")
         .withNetwork(network)
         .withEnv(jaegerEnvs)
-        .waitingFor(new BoundPortHttpWaitStrategy(16687).forStatusCode(204))
+        .waitingFor(new BoundPortHttpWaitStrategy(16687).forStatusCodeMatching(statusCode -> statusCode >= 200 && statusCode < 300))
         .withExposedPorts(16687, 16686);
     jaegerQuery.start();
 

--- a/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/JaegerElasticsearchEnvironment.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/JaegerElasticsearchEnvironment.java
@@ -48,7 +48,7 @@ public class JaegerElasticsearchEnvironment {
 
   public static String elasticsearchVersion() {
     String version = System.getProperty("elasticsearch.version", System.getenv("ELASTICSEARCH_VERSION"));
-    return version != null ? version : "5.6.16";
+    return version != null ? version : "5.6.9";
   }
 
   public void start(Map<String, String> jaegerEnvs, String jaegerVersion, String elasticsearchVersion) {

--- a/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/JaegerElasticsearchEnvironment.java
+++ b/jaeger-spark-dependencies-elasticsearch/src/test/java/io/jaegertracing/spark/dependencies/elastic/JaegerElasticsearchEnvironment.java
@@ -46,9 +46,14 @@ public class JaegerElasticsearchEnvironment {
   private String collectorUrl;
   private String zipkinCollectorUrl;
 
-  public void start(Map<String, String> jaegerEnvs, String jaegerVersion) {
+  public static String elasticsearchVersion() {
+    String version = System.getProperty("elasticsearch.version", System.getenv("ELASTICSEARCH_VERSION"));
+    return version != null ? version : "5.6.16";
+  }
+
+  public void start(Map<String, String> jaegerEnvs, String jaegerVersion, String elasticsearchVersion) {
     network = Network.newNetwork();
-    elasticsearch = new GenericContainer<>("docker.elastic.co/elasticsearch/elasticsearch:5.6.9")
+    elasticsearch = new GenericContainer<>(String.format("docker.elastic.co/elasticsearch/elasticsearch:%s", elasticsearchVersion))
         .withNetwork(network)
         .withNetworkAliases("elasticsearch")
         .waitingFor(new BoundPortHttpWaitStrategy(9200).forStatusCode(200))

--- a/jaeger-spark-dependencies/pom.xml
+++ b/jaeger-spark-dependencies/pom.xml
@@ -98,7 +98,7 @@
                   </includes>
                 </filter>
                 <filter>
-                  <artifact>net.jpountz.lz4:*</artifact>
+                  <artifact>org.lz4:*</artifact>
                   <includes>
                     <include>**</include>
                   </includes>
@@ -113,6 +113,13 @@
                 <filter>
                   <!-- elasticsearch -->
                   <artifact>commons-httpclient:commons-httpclient</artifact>
+                  <includes>
+                    <include>**</include>
+                  </includes>
+                </filter>
+                <filter>
+                  <!-- org.apache.xerces.jaxp.DocumentBuilderFactoryImpl -->
+                  <artifact>xerces:xercesImpl</artifact>
                   <includes>
                     <include>**</include>
                   </includes>

--- a/pom.xml
+++ b/pom.xml
@@ -37,8 +37,7 @@
     <project.build.resourceEncoding>UTF-8</project.build.resourceEncoding>
 
     <version.scala.binary>2.11</version.scala.binary>
-    <!-- spark-cassandra-connector doesn't yet support spark 2.2 -->
-    <version.org.apache.spark>2.1.1</version.org.apache.spark>
+    <version.org.apache.spark>2.3.0</version.org.apache.spark>
     <version.io.jaegertracing>0.28.0</version.io.jaegertracing>
     <version.io.opentracing>0.31.0</version.io.opentracing>
     <version.io.zipkin.brave-brave>4.6.0</version.io.zipkin.brave-brave>

--- a/pom.xml
+++ b/pom.xml
@@ -44,7 +44,7 @@
     <version.io.zipkin.reporter-zipkin-sender-okhttp3>1.0.2</version.io.zipkin.reporter-zipkin-sender-okhttp3>
     <version.junit>4.12</version.junit>
     <version.org.assertj>3.8.0</version.org.assertj>
-    <version.org.testcontainers>1.7.3</version.org.testcontainers>
+    <version.org.testcontainers>1.12.0</version.org.testcontainers>
     <version.com.squareup.okhttp3-okhttp>3.8.0</version.com.squareup.okhttp3-okhttp>
     <version.org.awaitility-awaitility>3.0.0</version.org.awaitility-awaitility>
 

